horovod/common/operations.cc

```diff
@@ -956,34 +958,60 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
                                      cudaMemcpyDeviceToDevice, stream))
           offset += e.tensor->size();
         }
-        buffer_len = (size_t)offset;
+
         if (timeline.Initialized() || horovod_global.ddl_initialized) {
           RECORD_EVENT(entries, event_queue, MEMCPY_IN_FUSION_BUFFER, stream)
         }
 
+        buffer_len = (size_t)offset;
+
         // Set the input data to originate from the buffer.
         fused_input_data = buffer_data;
 
         // Perform the reduction on the fusion buffer.
         for (auto& e : entries) {
           num_elements += e.tensor->shape().num_elements();
         }
+
       } else {
         fused_input_data = first_entry.tensor->data();
         buffer_data = (void*)first_entry.output->data();
         num_elements = first_entry.tensor->shape().num_elements();
         buffer_len = (size_t)first_entry.output->size();
+
         if (horovod_global.ddl_initialized) {
           // Copy input buffer content to output buffer
           // because DDL only supports in-place allreduce
           CUDA_CHECK(entries, "cudaMemcpyAsync",
-                     cudaMemcpyAsync(buffer_data, fused_input_data,
-                                     buffer_len,
+                     cudaMemcpyAsync(buffer_data, fused_input_data, buffer_len,
                                      cudaMemcpyDeviceToDevice, stream))
           RECORD_EVENT(entries, event_queue, MEMCPY_IN_FUSION_BUFFER, stream)
         }
       }
 
+      // For hierarchical allreduce pad the buffer to make the effective number
+      // of elements a multiple of local size
+      if (horovod_global.hierarchical_allreduce &&
+          num_elements % horovod_global.local_size != 0) {
+        int64_t num_elements_to_pad =
+            horovod_global.local_size -
+            (num_elements % horovod_global.local_size);
+        int64_t bytes_to_pad =
+            num_elements_to_pad * sizeof(first_entry.tensor->dtype());
+
+        void* buffer_data_at_offset = (uint8_t*)buffer_data + buffer_len;
+        void* fused_input_data_at_offset =
+            (uint8_t*)fused_input_data + buffer_len;
+
+        CUDA_CHECK(
+            entries, "cudaMalloc",
+            cudaMalloc((void**)&fused_input_data_at_offset, bytes_to_pad))
```

Comment:
You're right :) 

Since it is potentially wasteful to allocate new memory for the whole tensor at every call to `PerformOperation` whenever `entries.size()==1`, I feel like the only option is to break up the tensor into suitably sized pieces, which both of your suggestions already do.

I assume that by `memcpy` to the temporary buffer you meant always copying to the fusion buffer? If that is the case I feel closer to this solution, which seems cleaner and results in a slightly more unified way of dealing with every call to `PerformOperation` by eliminating some of the `if(entries.size()>1)` branching (though not completely). It also means that we impose a uniform upper bound on the data size of all allreduce operations through a single parameter (fusion threshold), which might potentially have other benefits.

Just to clarify, are you suggesting to initially allocate the fusion buffer with size `tensor_fusion_threshold+padding`, where `padding` is the small amount we need to add so that the sum is divisible by `hvd.local_size()`?

Reply:
this cudaMalloc default is 0 init?  and cudaMalloc cost time

Reply:
@alsrgv I ended up implementing the second (hybrid) approach since it required less change in the architecture. Please let me know what you think!

@anpark cudaMalloc did not solve the problem anyway so I removed it.